File: experiments/archive_me_legacy_flutter/apps/voicememory_mobile/lib/features/curiosity_loop/services/curiosity_hook_engine.dart
import '../models/curiosity_hook.dart';

/// Metadata extracted from a post-save voice moment for curiosity targeting.
class CuriosityHookEntryMetadata {
  const CuriosityHookEntryMetadata({
    required this.entryId,
    required this.createdAt,
    required this.extractedAnchors,
    this.emotionalTone,
    this.hasBlockers = false,
    this.entryCount = 1,
  });

  final String entryId;
  final DateTime createdAt;
  final List<String> extractedAnchors;
  final String? emotionalTone;
  final bool hasBlockers;
  final int entryCount;
}

abstract final class _CuriosityHookCopy {
  _CuriosityHookCopy._();

  static String blockerPrompt(String anchor) =>
      'Before "$anchor" showed up again, what got in the way?';

  static String momentumPrompt(String anchor) =>
      'You named "$anchor" — what felt different about it this time?';

  static String anchorFollowUpPrompt(String anchor) =>
      'Next time "$anchor" comes up, what do you want to notice first?';

  static String returnWatchPrompt(String anchor) =>
      'Come back when "$anchor" shows up again and say what changed.';
}

/// Builds targeted post-save curiosity hooks from entry metadata only.
abstract final class CuriosityHookEngine {
  CuriosityHookEngine._();

  static const _maxAnchorChars = 72;
  static const _maxRecentTypes = 4;

  static CuriosityHook? build({
    required CuriosityHookEntryMetadata metadata,
    Iterable<CuriosityHookType> recentHookTypes = const [],
    DateTime? now,
  }) {
    if (metadata.entryId.trim().isEmpty) return null;

    final anchor = _primaryAnchor(metadata.extractedAnchors);
    if (anchor == null) return null;

    final hookType = _selectHookType(
      metadata: metadata,
      recentHookTypes: recentHookTypes,
    );
    if (hookType == null) return null;

    final prompt = _promptFor(hookType: hookType, anchor: anchor);
    if (prompt.isEmpty) return null;

    final clock = now ?? DateTime.now().toUtc();
    return CuriosityHook(
      id: _hookId(metadata.entryId, clock),
      entryId: metadata.entryId,
      createdAt: metadata.createdAt.toUtc(),
      primaryAnchor: anchor,
      hookType: hookType,
      dynamicPrompt: prompt,
    );
  }

  static String? _primaryAnchor(List<String> extractedAnchors) {
    for (final raw in extractedAnchors) {
      final trimmed = raw.trim();
      if (trimmed.isEmpty) continue;
      if (trimmed.length <= _maxAnchorChars) return trimmed;
      return '${trimmed.substring(0, _maxAnchorChars - 1).trim()}…';
    }
    return null;
  }

  static CuriosityHookType? _selectHookType({
    required CuriosityHookEntryMetadata metadata,
    required Iterable<CuriosityHookType> recentHookTypes,
  }) {
    final recent = recentHookTypes.take(_maxRecentTypes).toList();
    final candidates = <CuriosityHookType>[];

    if (metadata.hasBlockers) {
      candidates.add(CuriosityHookType.blocker);
    }
    if (_hasMomentumTone(metadata.emotionalTone) && !metadata.hasBlockers) {
      candidates.add(CuriosityHookType.momentum);
    }
    if (metadata.entryCount >= 3) {
      candidates.add(CuriosityHookType.returnWatch);
    }
    candidates.add(CuriosityHookType.anchorFollowUp);

    for (final candidate in candidates) {
      if (!_recentlyUsed(candidate, recent)) return candidate;
    }

    for (final candidate in candidates) {
      if (recent.isEmpty || recent.first != candidate) return candidate;
    }
    return candidates.isEmpty ? null : candidates.first;
  }

  static bool _recentlyUsed(
    CuriosityHookType candidate,
    List<CuriosityHookType> recent,
  ) {
    if (recent.isEmpty) return false;
    if (recent.first == candidate) return true;
    if (recent.length >= 2 && recent[1] == candidate) return true;
    return false;
  }

  static bool _hasMomentumTone(String? emotionalTone) {
    final tone = emotionalTone?.trim().toLowerCase();
    if (tone == null || tone.isEmpty) return false;
    const momentumSignals = [
      'lighter',
      'lighter than',
      'hopeful',
      'calm',
      'relief',
      'clearer',
      'easier',
      'steady',
      'settled',
    ];
    return momentumSignals.any(tone.contains);
  }

  static String _promptFor({
    required CuriosityHookType hookType,
    required String anchor,
  }) {
    return switch (hookType) {
      CuriosityHookType.blocker => _CuriosityHookCopy.blockerPrompt(anchor),
      CuriosityHookType.momentum => _CuriosityHookCopy.momentumPrompt(anchor),
      CuriosityHookType.returnWatch => _CuriosityHookCopy.returnWatchPrompt(
        anchor,
      ),
      CuriosityHookType.anchorFollowUp =>
        _CuriosityHookCopy.anchorFollowUpPrompt(anchor),
    };
  }

  static String _hookId(String entryId, DateTime createdAt) =>
      'curiosity_${entryId}_${createdAt.millisecondsSinceEpoch}';
}

File: apps/api/src/services/loops/curiosity_hook_engine.ts
import "server-only";

import type { CuriosityHook, CuriosityHookEntryMetadata, CuriosityHookType } from "./types";

const MAX_ANCHOR_CHARS = 72;
const MAX_RECENT_TYPES = 4;

const MOMENTUM_TONE_SIGNALS = [
  "lighter",
  "lighter than",
  "hopeful",
  "calm",
  "relief",
  "clearer",
  "easier",
  "steady",
  "settled",
];

/**
 * Selects hook targeting metadata only — no notification copy lives here.
 * Message text is built separately from `fact_ledger` evidence.
 */
export function buildCuriosityHook(input: {
  metadata: CuriosityHookEntryMetadata;
  recentHookTypes?: readonly CuriosityHookType[];
  now?: Date;
}): CuriosityHook | null {
  const entryId = input.metadata.entryId.trim();
  if (!entryId) return null;

  const anchor = primaryAnchor(input.metadata.extractedAnchors);
  if (!anchor) return null;

  const hookType = selectHookType(
    input.metadata,
    input.recentHookTypes ?? [],
  );
  if (!hookType) return null;

  const createdAt = input.metadata.createdAt.trim();
  if (!createdAt || !Number.isFinite(Date.parse(createdAt))) return null;

  const clock = input.now ?? new Date();
  return {
    id: hookId(entryId, clock),
    entryId,
    createdAt: new Date(createdAt).toISOString(),
    primaryAnchor: anchor,
    hookType,
    sourceEntryId: entryId,
  };
}

function primaryAnchor(extractedAnchors: readonly string[]): string | null {
  for (const raw of extractedAnchors) {
    const trimmed = raw.trim();
    if (!trimmed) continue;
    if (trimmed.length <= MAX_ANCHOR_CHARS) return trimmed;
    return `${trimmed.slice(0, MAX_ANCHOR_CHARS - 1).trim()}…`;
  }
  return null;
}

function selectHookType(
  metadata: CuriosityHookEntryMetadata,
  recentHookTypes: readonly CuriosityHookType[],
): CuriosityHookType | null {
  const recent = recentHookTypes.slice(0, MAX_RECENT_TYPES);
  const candidates: CuriosityHookType[] = [];

  if (metadata.hasBlockers) candidates.push("blocker");
  if (hasMomentumTone(metadata.emotionalTone) && !metadata.hasBlockers) {
    candidates.push("momentum");
  }
  if ((metadata.entryCount ?? 0) >= 3) candidates.push("returnWatch");
  candidates.push("anchorFollowUp");

  for (const candidate of candidates) {
    if (!recentlyUsed(candidate, recent)) return candidate;
  }

  for (const candidate of candidates) {
    if (recent.length === 0 || recent[0] !== candidate) return candidate;
  }

  return candidates[0] ?? null;
}

function recentlyUsed(
  candidate: CuriosityHookType,
  recent: readonly CuriosityHookType[],
): boolean {
  if (recent.length === 0) return false;
  if (recent[0] === candidate) return true;
  if (recent.length >= 2 && recent[1] === candidate) return true;
  return false;
}

function hasMomentumTone(emotionalTone: string | undefined): boolean {
  const tone = emotionalTone?.trim().toLowerCase();
  if (!tone) return false;
  return MOMENTUM_TONE_SIGNALS.some((signal) => tone.includes(signal));
}

function hookId(entryId: string, createdAt: Date): string {
  return `curiosity_${entryId}_${createdAt.getTime()}`;
}
